tests/std/tests/P0769R2_shift_left_shift_right/test.cpp

```diff
@@ -7,6 +7,16 @@
 #include <iterator>
 #include <list>
 #include <vector>
+#include <version>
+
+#if __cpp_lib_shift >= 202202L
+// Test ranges::shift_left, ranges::shift_right from P2440R1
+//#define TEST_RANGES
+/*
+ * Disabled until `forward_list::_Unchecked_end` is fixed to return the same type as `forward_list::end()._Unwrapped()`
+ * to work with `_Uend` in <xutility>

```

Comment:
Is this something that can be fixed while preserving ABI? Or does it indicate an overly strong assumption in our unwrapping machinery (e.g. similar to #2591)?

Reply:
The unwrapping machinery assumes that `_Get_unwrapped(ranges::end(r))` and `_Uend(ranges::end(r))` are equivalent expressions (`_Uend` actually asserts that they have the same type). This isn't true for `forward_list`: `_Unchecked_end()` yields a sentinel (for nice codegen properties) but `end()` has to return an iterator because the Standard says so.

I removed the `static_assert` from `_Uend`, and taught the range overload of `_Get_final_iterator` that `_Uend` for a `common_range` might not return an iterator, in which case it needs to `_Get_unwrapped(end())` instead of returning `_Unchecked_end()`. (I recall no other places that need similar cleanup.)